feat: adding attachments to snippets.

lorenzo132 opened the pull request:
This adds the ability to add attachments to snippets.

lorenzo132 committed:
feat: adding attachments to snippets.

This adds the ability to add attachments to snippets

lorenzo132 committed:
black formatting

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -343,14 +389,122 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+                    view.stop()
+
+                confirm_button = discord.ui.Button(label="✓ Confirm", style=discord.ButtonStyle.green)
+                cancel_button = discord.ui.Button(label="✗ Cancel", style=discord.ButtonStyle.red)
```
Are these symbols consistent with the confirm and cancel across the bot? I feel like usually it is an emoji or just no symbol.

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -343,14 +389,122 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+
+                embed = discord.Embed(
+                    title="Confirm Large Attachment",
+                    description=f"The attachment is {attachment.size / (1024 * 1024):.2f} MB (≥2 MB).\n"
```
the 1024 1024 makes this a MiB not MB.

lorenzo132 committed:
Merge branch 'development' into users/lorenzo/snippet-attachment-support

lorenzo132 committed:
Merge branch 'development' into users/lorenzo/snippet-attachment-support

martinbndr reviewed (changes requested):
<img width="815" height="689" alt="image" src="https://github.com/user-attachments/assets/3775720b-ff81-4fbc-bc2a-d5c4fbbcb6a5" />

Is it intented, that an image is sent seperate comparing to the normal reply? If not maybe we should consider letting images being added to the embed aswell to have it consistent to the replying?

lorenzo132 committed:
change displaying method.

lorenzo132 committed:
black formatting

lorenzo132 committed:
Merge branch 'development' into users/lorenzo/snippet-attachment-support

lorenzo132 committed:
Merge branch 'development' into users/lorenzo/snippet-attachment-support

lorenzo132 committed:
Merge branch 'development' into users/lorenzo/snippet-attachment-support

lorenzo132 committed:
Fix snippet attachment handling

lorenzo132 committed:
Preserve forwarded attachments with snippets

lorenzo132 committed:
fix: help command

lorenzo132 committed:
Merge branch 'development' into users/lorenzo/snippet-attachment-support

StephenDaDev reviewed (approved):
tested working, no serious issues found

sebkuip commented on cogs/modmail.py:
```diff
@@ -250,10 +274,17 @@ async def snippet(self, ctx, *, name: str.lower = None):
+                snippet_text = self._get_snippet_text(snippet_data)
+                has_attachment = self._has_snippet_attachment(snippet_data)
+
+                description = snippet_text if snippet_text else "(No text content)"
```
If a snippet is only an attachment, having it show up as `(No text content) This snippet has an attachment` seems weird. I would put it as something like `this is an attachment-only snippet`

sebkuip commented on cogs/modmail.py:
```diff
@@ -274,10 +305,15 @@ async def snippet(self, ctx, *, name: str.lower = None):
+
+            display_value = return_or_truncate(snippet_text, 350) if snippet_text else "(No text)"
+            if has_attachment:
+                display_value = "📎 " + display_value
```
Maybe make the embed have a footer text explaining that the paperclip emoji means the embed has an attachment?

sebkuip commented on cogs/modmail.py:
```diff
@@ -250,10 +274,17 @@ async def snippet(self, ctx, *, name: str.lower = None):
+
+                description = snippet_text if snippet_text else "(No text content)"
+                if has_attachment:
+                    description += "\n\n📎 *This snippet has an attachment.*"
```
I'd say show the attachment when you request the snippet. In the list it's fine to leave out but if you just pull up that snippet it should show.

sebkuip commented on cogs/modmail.py:
````diff
@@ -292,20 +328,28 @@ async def snippet_raw(self, ctx, *, name: str.lower):
+            val = truncate(escape_code_block(snippet_text), 2048 - 7) if snippet_text else "(No text content)"
+            description = f"```\n{val}```"
+            if has_attachment:
+                description += "\n\n📎 *This snippet has an attachment.*"
````
Similar to the regular version of this command, show the attachment so the staff member can see what it sends.

sebkuip commented on cogs/modmail.py:
```diff
@@ -347,14 +393,128 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+        file_id = None
+        attachment_info = None
+        if ctx.message.attachments:
+            attachment = ctx.message.attachments[0]
```
Don't silently ignore multiple attachments. Either allow it, or give the user an error that only a single attachment is allowed.

sebkuip commented on cogs/modmail.py:
```diff
@@ -347,14 +393,128 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+
+            # Validate file size
+            max_size_mb = self.bot.config.get("snippet_attachment_max_size")
+            max_size_bytes = max_size_mb * 1024 * 1024
```
Wouldn't `max_size_mb * 1024 ** 2` be cleaner? Possibly `max_size_mb * (1024 ** 2)` to make it more clear?
Also add a short comment to explain why this calculation is done.

StephenDaDev replied:
on L363 you are informing users that you have a 10MB limit but you are calculating in MiB instead

sebkuip commented on cogs/modmail.py:
```diff
@@ -347,14 +393,128 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+
+            # Confirmation for attachments 2MB or higher
+            confirm_msg = None
+            if attachment.size >= 2 * 1024 * 1024:
```
Similar as comment above

martinbndr reviewed (dismissed):
Looks good.

sebkuip commented on cogs/modmail.py:
```diff
@@ -347,14 +393,128 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+                return await ctx.send(embed=embed)
+
+        # Require at least text or attachment
+        if not value and not file_id:
```
This if statement acts as a guard, and should be at the top of the method. Probably change it to `if not value and len(message.attachments) == 0:`

sebkuip commented on cogs/modmail.py:
```diff
@@ -457,28 +620,114 @@ async def snippet_remove(self, ctx, *, name: str.lower):
             self.bot.snippets.pop(name)
             await self.bot.config.update()
+            if file_id and not await self.bot.api.delete_snippet_attachment(file_id):
+                logger.warning("Failed to delete snippet attachment for %s.", name)
```
This should either inform the user in chat, or fail the command entirely so you don't end up with tons of data on gridFS when it errors a few times.

sebkuip commented on cogs/modmail.py:
````diff
@@ -457,28 +620,114 @@ async def snippet_remove(self, ctx, *, name: str.lower):
         ```
+
+        Attach a new file to replace the existing attachment.
+        Provide text without attachment to keep the existing attachment.
````
I feel like an edit should be able to remove the attachment as well. Isn't it better that providing no attachment deletes it, instead of having the user needing to delete the snippet if they want the attachment gone?

StephenDaDev replied:
agree

sebkuip commented on cogs/utility.py:
```diff
@@ -109,7 +109,7 @@ async def format_cog_help(self, cog, *, no_cog=False):
 
     def process_help_msg(self, help_: str):
-        return help_.format(prefix=self.context.clean_prefix) if help_ else "No help message."
+        return help_.replace("{prefix}", self.context.clean_prefix) if help_ else "No help message."
```
Why is this randomly edited? This is out of scope for the PR

lorenzo132 replied:
This was to fix the help command, it was introduced somewhere, and it was a bug i found was small enough to include here.

sebkuip commented on core/config.py:
```diff
@@ -422,6 +426,14 @@ def get(self, key: str, *, convert: bool = True) -> typing.Any:
+        elif key in self.megabytes:
+            if not isinstance(value, int):
+                try:
+                    value = int(value)
```
Might also want to guard against negative values. Also this causes rounding when a comma value is entered, is that intentional?

sebkuip reviewed (dismissed):
Review for now. Did not check all, and is purely a code review. Have to do some testing later myself and look further into the code.

StephenDaDev commented on cogs/modmail.py:
````diff
@@ -315,6 +359,8 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
         {prefix}snippet add "two word" this is a two word snippet.
         ```
+
+        You can also attach a file (max 10 MB) to include with the snippet.
````
I think this needs to have a disclaimer that explains that this is storage intensive and may cause issues if used frequently.

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -347,14 +393,128 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+                embed = discord.Embed(
+                    title="Error",
+                    color=self.bot.error_color,
+                    description=f"Attachment exceeds the maximum file size of {max_size_mb} MB. "
```
same as above

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -347,14 +393,128 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+                        )
+                    confirmed = False
+                    await interaction.response.edit_message(
+                        content="❌ Cancelled. Snippet not created.", view=None, embed=None
```
I think the use of emojis does not align with informal styling guidelines of the bot, I can't think of any other cases where they are used in this context

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -347,14 +393,128 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+                    )
+                    view.stop()
+
+                confirm_button = discord.ui.Button(label="✓ Confirm", style=discord.ButtonStyle.green)
```
same here, as far as I can think of, there is no use of symbols

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -347,14 +393,128 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+                if confirmed is None or not confirmed:
+                    if confirmed is None:
+                        await confirm_msg.edit(
+                            content="⏱️ Timed out. Snippet not created.", view=None, embed=None
```
same

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -347,14 +393,128 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+            embed = discord.Embed(
+                title="Error",
+                color=self.bot.error_color,
+                description="You must provide either text content or an attachment for the snippet.",
```
Usually our current convention for misformed is to send the help menu again

<img width="459" height="324" alt="Image" src="https://github.com/user-attachments/assets/a7a040bf-e9df-48d9-a686-9fde4bc8ec7e" />

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -457,28 +620,114 @@ async def snippet_remove(self, ctx, *, name: str.lower):
+                    embed = discord.Embed(
+                        title="Error",
+                        color=self.bot.error_color,
+                        description=f"Attachment exceeds the maximum file size of {max_size_mb} MB. "
```
It seems like if adding to snippet via an edit the large file warning is not displayed (logic in R449-R453).

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -457,28 +620,114 @@ async def snippet_remove(self, ctx, *, name: str.lower):
+            if value:
+                description += f'\nText: "{truncate(value, 100)}"'
+            if attachment_info:
+                description += f"\n📎 New attachment: `{attachment_info}`"
```
same comments on emoji use

StephenDaDev commented on core/config_help.json:
```diff
@@ -856,6 +856,18 @@
     ]
   },
+  "snippet_attachment_max_size": {
+    "default": "10 (MB)",
```
same comments on MB vs MiB

StephenDaDev commented on core/config_help.json:
```diff
@@ -856,6 +856,18 @@
+    ],
+    "notes": [
+      "Attachments larger than this size will be rejected when adding or editing snippets.",
+      "Value is specified in megabytes (MB)."
```
I think a note explaining the cons of raising this filesize without considering long term implications should be added

StephenDaDev reviewed (changes requested):
please see my mostly formatting consistency comments

lorenzo132 committed:
Resolve feedback.

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -116,6 +119,144 @@ def _resolve_user(self, user_str):
+            nonlocal confirmed
+            if interaction.user.id != ctx.author.id:
+                return await interaction.response.send_message(
+                    "Only the command author can cancel.", ephemeral=True
```
I think this should be an embed to be consistent with the rest of the bot, using error color

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -116,6 +119,144 @@ def _resolve_user(self, user_str):
+                )
+            confirmed = False
+            await interaction.response.edit_message(
+                content=f"Cancelled. Snippet not {past_tense}.", view=None, embed=None
```
same here

StephenDaDev commented on cogs/modmail.py:
```diff
@@ -347,15 +541,85 @@ async def snippet_add(self, ctx, name: str.lower, *, value: commands.clean_conte
+                "Please try again with another name."
+            )
+            if cleanup_failed:
+                description += " The uploaded file could not be cleaned up."
```
how should users handle this when this happens?

lorenzo132 committed:
Resolve requests.

lorenzo132 committed:
fix broken attachment state

This solves that if conversion later fails, the embed is sent without its referenced file.

StephenDaDev reviewed (dismissed):
Tested working, no further comments from me

lorenzo132 committed:
Merge branch 'development' into users/lorenzo/snippet-attachment-support

sebkuip commented on cogs/modmail.py:
```diff
@@ -250,13 +404,25 @@ async def snippet(self, ctx, *, name: str.lower = None):
+                    description = snippet_text
+                elif has_attachment:
+                    description = "This is an attachment-only snippet."
+                else:
```
In what case would this else be reached? A snippet must have an attachment or text right? It shouldn't be possible to make a snippet with no content.

sebkuip commented on core/thread.py:
```diff
@@ -2078,6 +2109,26 @@ def lottie_to_png(data):
+
+            if not embedded_image:
+                embed.set_image(url=f"attachment://{a.filename}")
+                embed.add_field(name="Image", value=a.filename)
```
This is inconsistent with how reply works, where the filename is a link (top is snippet, bottom is reply)

<img width="437" height="852" alt="Image" src="https://github.com/user-attachments/assets/e9c0d791-e719-40bd-8890-8de1e247ddae" />

sebkuip reviewed (changes requested):
There are some issues still in the code. Please resolve these
